### ISSUE TITLE
 [Flang] Incorrect diagnose on NULL(mold) actual to allocatable dummy

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#115984]
Consider the following code:
```
module m
    type Base
        integer i
    end type

    type, extends(Base) :: Child
        integer j
    end type
end module

program argAssociation001
use m
    type(Child), allocatable :: c

    allocate(c, SOURCE=Child(2,-2))
    call sub1(null(c))

    contains
    subroutine sub1(arg)
        type(Child), allocatable :: arg
    end subroutine
end
```

Flang issues an error as
```
t2.f:16:15: error: A null pointer may not be associated with allocatable dummy argument 'arg=' without INTENT(IN)
      call sub1(null(c))
```

The code seems conforming to me as I can't find such constraint in the standard. 
Both XLF and ifort compile the code successfully.

**Comment by klausler** [comment]
That is a valid error.  You can pass NULL as an actual argument when the corresponding dummy is a pointer, is optional, or is an INTENT(IN) allocatable.  But it doesn't make sense to pass NULL for a dummy that is allocatable and neither optional nor INTENT(IN) -- there's no effective actual argument with which the dummy argument can be associated.

**Comment by DanielCChen** [comment]
@klausler Reopen for some clarification. 
The standard says the following about the dummy argument.

```
The actual argument shall be allocatable. It is permissible for the actual argument to have an allocation status of unallocated.
```

For intrinsic NULL(MOLD).
```
Result Characteristics. If MOLD is present, the characteristics are the same as MOLD.
...
Result: The result is a disassociated pointer or an unallocated allocatable entity.
```

This is exactly this test case is testing. The actual argument is an unallocated allocatable object that corresponds to an allocatable dummy. 

It initializes the dummy argument using NULL() intrinsic rather than deallocate `c` first then pass it.

**Comment by klausler** [comment]
But `NULL(MOLD=c)` is an expression, not a variable, so it is not possible for the actual argument to become defined.  That's why I require the `INTENT(IN)`.  I should probably relax that to also allow this case of a dummy argument with no `INTENT`.

**Comment by DanielCChen** [comment]
@klausler Thanks for the fix.
I still think the "warning" message is bit confusing. 

`NULL(MOLD)` is a unique case in the sense that the result is an `... unallocated allocatable entity` according to the standard unlike a user defined function that returns an allocatable result. User function allocatable result is the value of it, and it doesn't have the allocatable attribute. 

The message requires the dummy argument to have the `INTENT(IN)` attribute, which it makes the dummy argument useless as users cannot change its allocation status, which is unallocated.

Same for the case of `call sub1(NULL())` without the `MOLD` argument as the standard says it would have the TKR of the corresponding dummy argument. 

I think this is a legal case of using intrinsic `NULL`.

**Comment by klausler** [comment]
But it is not at all clear that `NULL(MOLD=allocatable)` produces a *definable variable*.  It doesn't work in XLF as an (unused) input list data item or as a variable in an ALLOCATE or DEALLOCATE statement, and if it appears as the selector of ASSOCIATE, it's not definable.

It is indeed allowed by (only) XLF as the left-hand side of an intrinsic assignment statement and as an actual argument association with an allocatable dummy argument.

So is XLF's behavior something that you believe to be conforming?  What are the contexts in which `NULL(a)` *can* be used as a definable variable, and what is the lifetime of the anonymous allocatable entity?

**Comment by DanielCChen** [comment]
Right. 
I don't think allowing `NULL(C) = 4` is intentional as it should fall into the same category as the ALLOCATE and DEALLOCATE statement that require a variable (NULL(C) when c being an allocatable is definitely not a variable). 

I now think the wording in the standard that describes the result of `NULL(MOLD)` being `an unallocated allocatable entity` is misleading. 
I also agree now that this should be an error condition that shall have the same treatment as the following code.
```
interface
  function foo()
    real, allocatable :: foo
  end function
end interface
call sub(foo())
contains
  subroutine sub(arg)
    real, allocatable :: arg
  end
end
``` 

Both Flang and XLF issued a similar error message  as
```
6:10: error: ALLOCATABLE dummy argument 'arg=' must be associated with an ALLOCATABLE actual argument
  call sub(foo())
           ^^^^^
```

I think issuing this error message is probably the better fix for this issue.

**Comment by klausler** [comment]
No, there's a distinction to be made between `NULL(MOLD=allocatable)` and a function whose result variable is allocatable.  The result of `NULL(MOLD=allocatable)` is explicitly defined to be an allocatable, whereas the result (in the caller) of the function is explicitly defined to *not* have the allocatable attribute.

**Comment by DanielCChen** [comment]
That is exactly where I think the confusion is coming from. 
The standard is explicit that only the variables and components can have the ALLOCATABLE attribute.
If `NULL(MOLD=allocatable)` is allocatable (have the ALLOCATABLE attribute), then it should be allowed at any places an allocatable variable can appear. However, it doesn't make sense for the cases you listed at the above. For instance, LHS of an intrinsic assignment, ALLOCATE statement.
Passing it as an actual argument that corresponds to an allocatable dummy is not as obvious as the other usages, but it should be consistent.

**Comment by klausler** [comment]
The description of `NULL(MOLD=allocatable)` explicitly says that its result is an "unallocated allocatable entity".  I agree that only named variables and components can have the ALLOCATABLE attribute, but perhaps it is possible for something to be an allocatable without having the attribute.

There must be some justification for `NULL(MOLD=allocatable)` to have been defined in this way -- there must be some context in which it may appear.  The only one that I can come up with is as an actual argument for a dummy allocatable that is not required to be definable.

**Comment by DanielCChen** [comment]
Ok. I will try to get some answers from the committee.

**Comment by DanielCChen** [comment]
@klausler The code is indeed valid. The response from the standard committee is that it is similar to passing a literal `1.0` to a real dummy argument. It is not definable, so the invalid intent would be INTENT(OUT) and INTENT(INOUT).
It also acknowledged that there is an issue with the wording in the standard in terms of what is allocatable. There is a suggestion to separate variables from the result of `NULL(allocatable)` so that both can be independently allocatable, but that would be for future consideration.

**Comment by klausler** [comment]
Do you have a link to this response, or are you able to quote it here?

**Comment by akuhlens** [comment]
I think there may be another wrinkle in the standard around this:

> Only variables and components can have the ALLOCATABLE attribute. The result of referencing a function whose result variable has the ALLOCATABLE attribute is a value that does not itself have the ALLOCATABLE attribute. Note in 8.5.3

> If an executable construct references a function whose result is allocatable or has an allocatable subobject, and the function reference is executed, an allocatable result and any allocated allocatable subobject of the result is deallocated after execution of the innermost executable construct containing the reference. - 9.7.3.2.4

> The characteristics of a function result are its declared type, type parameters, rank, whether it is polymorphic, whether it is allocatable 15.3.3.1

> allocatable = having the ALLOCATABLE attribute 3.2

The first and the last coexisting with either of the middle two quote seems contradictory to me. Either allocatable means something other than having the allocatable attribute (maybe there are multiple meanings meant) or results have the allocatable attribute and that needs some interpretation.

**Comment by jeanPerier** [comment]
> The first and the last coexisting with either of the middle two quote seems contradictory to me. Either allocatable means something other than having the allocatable attribute (maybe there are multiple meanings meant) or results have the allocatable attribute and that needs some interpretation.

It is probably hard to find good wording for the standard, but the intent is rather clear from the note and [this J3 thread](https://mailman.j3-fortran.org/pipermail/j3/2025-February/015131.html).

The conceptual idea is that there is a difference between the result variable, and the value produced by a function. An allocatable function does not produces a variable (you cannot assign to it), but a value. Conceptually, the result is an ALLOCATABLE variable during the call only, and on the caller side, it is just an expression value. To avoid having to make a copy of the result value on return for the sake of finalization, the standard extended the lifetime of the result variable to the use of its value, which adds a bit of confusion.

NULL is an intrinsic procedure, and intrinsic procedure specifications can override the specifications of user procedures. So these points do not completely apply here, although the discussions in this bug do show that, as Daniel mentioned [above](https://github.com/llvm/llvm-project/issues/115984#issuecomment-2637729483),  some wording lacks to clarify the specificity of NULL (producing a value that is allocatable, while as you pointed out, allocatable are defined to be variables).

**Comment by DanielCChen** [comment]
Right. at the moment, an allocatable entity can only be variable or component as @akuhlens pointed out. However, the code in this issue is intended to be legal, which requires some modification in the standard to allow an non-variable allocatable entity (limited to reference to `NULL(allocatable)` only).

**Comment by akuhlens** [comment]
I am going to take a look at this and probably downgrade the error to either a warning or no message at all in the near future.

**Comment by klausler** [comment]
reassigning to myself, as I already have a fix pending (see above) that I will clean up and merge.

I think now that it should not be an error to pass `NULL([allocatable])` as an actual argument corresponding to a dummy argument without intent, because it certainly is not an error to pass a structure constructor containing an allocatable component.
